src/state.rs

```diff
@@ -0,0 +1,58 @@
+//! Program state accounts
+
+use {
+    solana_program::clock::UnixTimestamp, solana_program::pubkey::Pubkey,
+    spl_pod::optional_keys::OptionalNonZeroPubkey,
+};
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum AccountType {
+    Uninitialized,
+    Config,
+    Stake,
+}
+
+/// Configuration for a staking system
+pub struct Config {
+    /// After a deactivation, defines the number of seconds that must pass before
+    /// the stake is inactive and able to be withdrawn
+    cooldown_time_secs: u64,
+    /// The maximum proportion that can be deactivated at once, given as basis
+    /// points (1 / 10,000)
+    max_deactivation_basis_points: u16,
+    /// Total number of tokens delegated to the system. Since anyone can transfer
+    /// tokens into the vault without passing through the program, this number
+    /// is maintained independently.
+    token_amount_delegated: u64,
+    /// Authority that can modify any elements in the config
+    config_authority: OptionalNonZeroPubkey,
+    /// Authority that can slash any stake
+    slash_authority: OptionalNonZeroPubkey,
+    /// Token account storing all stake
+    vault_token_account: Pubkey,
```

Comment:
Should this be the `mint` instead? The token account could have a deterministic derivation (e.g. , an ATA) and we probably need to validate the mint on stake instructions.

Reply:
This is just meant to be the token account -- but yeah, we'll need to do some validation on it during init, ie. to make sure the account is owned by the proper authority and doesn't have any extra extensions on it. And init also passes in the mint to do any checks on it too.